raiden/waiting.py

```diff
@@ -378,38 +385,76 @@ def wait_for_healthy(
     wait_for_network_state(raiden, node_address, NODE_NETWORK_REACHABLE, retry_timeout)
 
 
-def wait_for_transfer_success(
+class TransferWaitResult(Enum):
+    SECRET_REGISTERED_ONCHAIN = "secret registered onchain"
+    UNLOCKED = "unlocked"
+    UNLOCK_FAILED = "unlock_failed"
+
+
+def wait_for_received_transfer_result(
     raiden: "RaidenService",
     payment_identifier: PaymentID,
     amount: PaymentAmount,
     retry_timeout: float,
-) -> None:  # pragma: no unittest
-    """Wait until a transfer with a specific identifier and amount
-    is seen in the WAL.
+    secrethash: SecretHash,
+) -> TransferWaitResult:  # pragma: no unittest
+    """Wait for the result of a transfer with the specified identifier
+    and/or secrethash. Possible results are onchain secret registration,
+    successful unlock and failed unlock. For a successful unlock, the
+    amount is also checked.
 
     Note:
         This does not time out, use gevent.Timeout.
     """
     log_details = {"payment_identifier": payment_identifier, "amount": amount}
-    found = False
-    while not found:
+    result = None
+    while result is None:
         assert raiden, TRANSPORT_ERROR_MSG
         assert raiden.wal, TRANSPORT_ERROR_MSG
         assert raiden.transport, TRANSPORT_ERROR_MSG
 
         state_events = raiden.wal.storage.get_events()
         for event in state_events:
-            found = (
+            unlocked = (
                 isinstance(event, EventPaymentReceivedSuccess)
                 and event.identifier == payment_identifier
                 and event.amount == amount
             )
-            if found:
+            if unlocked:
+                result = TransferWaitResult.UNLOCKED
+                break
+            claim_failed = (
+                isinstance(event, EventUnlockClaimFailed)
+                and event.identifier == payment_identifier
+                and event.secrethash == secrethash
+            )
+            if claim_failed:
+                result = TransferWaitResult.UNLOCK_FAILED
                 break
 
-        log.debug("wait_for_transfer_success", **log_details)
+        state_changes = raiden.wal.storage.get_state_changes()
+        for state_change in state_changes:
+            registered_onchain = (
+                isinstance(state_change, ContractReceiveSecretReveal)
+                and state_change.secrethash == secrethash
+            )
+            if registered_onchain:
+                state_change_with_transfer = get_state_change_with_transfer_by_secrethash(
+                    raiden.wal.storage, secrethash
+                )
+                msg = "Expected ActionInitMediator/ActionInitTarget not found in state changes."
+                assert state_change_with_transfer is not None, msg
```

Comment:
This can be replaced by `assert isinstance(state_change, (ActionInitMediator, ActionInitTarget))`, this will satisfy the same check while adding information for mypy, removing the need for `type: ignore` below.